### DISCUSSION TITLE
Raw Generator in LazyCollection::make() is not lazy

### DISCUSSION THREAD

**Discussion by sebsel** [laravel/framework#33297] [General] [answered]
- Laravel Version: 7.15.0
- PHP Version: 7.4.0

### Description:

When using Lazy Collections, one can use the following to read a large file without loading it all in memory:

```php
use Illuminate\Support\LazyCollection;

LazyCollection::make(function () {
    $handle = fopen('file.txt', 'r');
    while ($line = fgets($handle)) yield $line;
})->map(...);
```

However, if you have a raw Generator (the result of calling a Generator function), then the LazyCollection will not handle it lazy anymore. The following will crash on out of memory when the file is too large:

```php
use Illuminate\Support\LazyCollection;

class FileReader
{
    public function read($file)
    {
        $handle = fopen($file, 'r');
        while ($line = fgets($handle)) yield $line;
    }
}

$reader = new FileReader();

LazyCollection::make($reader->read('file.txt'))->map(...);
```

The solution is to change it to the following:

```php
LazyCollection::make(fn() => yield from $reader->read('file.txt'))->map(...);
```

It cost me some time to figure out this bug, and it surprised me: I think Laravel can handle this for me?

### How it works

`LazyCollection::make()` calls the constructor, which calls `getArrayableItems()` from the `EnumeratesValues` trait.

https://github.com/laravel/framework/blob/a00484aec9fc2f66645c49e40ddc7ef9f13f0ecd/src/Illuminate/Support/LazyCollection.php#L29-L39

The generator falls into the case that checks for `Traversable`, which makes that Laravel calls `iterator_to_array()` on it. This will run the full generator, and thus in my example load the full file in memory.

https://github.com/laravel/framework/blob/a00484aec9fc2f66645c49e40ddc7ef9f13f0ecd/src/Illuminate/Support/Traits/EnumeratesValues.php#L870-L887

A solution could be to change the constructor of `LazyCollection` to the following:

```php
    public function __construct($source = null)
    {
        if ($source instanceof Closure || $source instanceof self) {
            $this->source = $source;
        } elseif ($source instanceof Generator) {
            $this->source = function () use ($source) {
                yield from $source;
            };
        } elseif (is_null($source)) {
            $this->source = static::empty();
        } else {
            $this->source = $this->getArrayableItems($source);
        }
    }

**Comment by driesvints** [comment]
Heya. This sounds like something more for discussions so I'm moving it there.

**Comment by driesvints** [comment]
Also pinging @JosephSilber since he probably has some insights here.

**Comment by JosephSilber** [comment, marked as answer]
The documentation does not say nor suggest that you can pass a `Generator` to the `LazyCollection`'s `make` method.

But [this keeps on coming up](https://github.com/laravel/ideas/issues/1880) every now and then, so maybe we should document it properly somewhere.

---

> I think Laravel can handle this for me?

I don't think it can, since iterators (AKA `Generator`s) aren't rewindable.

---

With your proposed solution:

```php
public function __construct($source = null)
{
    if ($source instanceof Generator) {
        $this->source = function () use ($source) {
            yield from $source;
        };
    }
    // other options omitted for brevity...
}
```

...this code will break:

```php
$collection->first() == $collection->first(); // false :(
```
